aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -501,6 +501,119 @@ Tensor norm(const Tensor& self, Scalar p) {
   return at::native::_norm(self, p);
 }
 
+static Tensor& _norm_matrix_out(Tensor &result, const Tensor &self, optional<Scalar> opt_p,
+                               IntArrayRef dim, bool keepdim, optional<ScalarType> opt_dtype) {
+  auto p = opt_p.value_or(2.0).toDouble();
+  TORCH_CHECK(self.device().type() == DeviceType::CPU || self.device().type() == DeviceType::CUDA,
+              "_norm_matrix only supports CPU AND CUDA device type, got: ", self.device().type());
+  TORCH_CHECK(self.layout() == Layout::Strided,
+              "_norm_matrix only supports strided layout, got: ", self.layout());
+  if ((dim.size() == 0) && (self.dim() == 2)) {
+    dim = {0, 1};
+  }
+  TORCH_CHECK(dim.size() == 2, "_norm_matrix: 'dim' must either specify 2 dimensions, or if ",
+    "'self' is 2-D 'dim' can specify 0 dimensions for a full reduction. Got 'dim' specifying ",
+    dim.size(), " dims and 'self' is ", self.dim(), "-D");
+  ScalarType scalarType = opt_dtype.has_value() ? opt_dtype.value() : self.scalar_type();
+  TORCH_CHECK(
+      at::isFloatingType(scalarType) || at::isComplexType(scalarType),
+      "Can only calculate the mean of floating types. Got ",
+      toString(scalarType),
+      " instead.");
+
+  auto dim_ = dim.vec();
+  maybe_wrap_dims(dim_, self.dim());
+  TORCH_CHECK(dim_[0] != dim_[1],
+    "Expected dims to be different, got (", dim[0], ", ", dim[1], ") instead");
+
+  Tensor result_;
+  Tensor self_;
+    
+  if (opt_dtype.has_value()) {
+    self_ = self.to(scalarType);
+  } else {
+    self_ = self;
+  }
+
+  if (std::abs(p) == 2) {
+    // Need to shift the reduction dims to the back, because at::svd will only operate on
+    // the last 2 dimensions
+    std::vector<int64_t> permuted_dims(self_.dim());
+    int64_t cur_permuted_dim = 0;
+    for (int64_t dim_ind = 0; dim_ind < self_.dim(); dim_ind++) {
+      if ((dim_ind != dim_[0]) && (dim_ind != dim_[1])) {
+        permuted_dims[cur_permuted_dim++] = dim_ind;
+      }
+    }
+    permuted_dims[cur_permuted_dim++] = dim_[0];
+    permuted_dims[cur_permuted_dim] = dim_[1];
+
+    result_ = std::get<1>(self_.permute(permuted_dims).svd()).abs();
+
+    if (p > 0) {
+      result_ = std::get<0>(result_.max(-1, keepdim));
+    } else {
+      result_ = std::get<0>(result_.min(-1, keepdim));
+    }
```

Comment:
Are you sure that `keepdim` will insert ones into right positions of the original, non-permuted sizes?

Reply:
No, I'm not sure. Making tests for the `keepdim=True` case is on my to-do list.

Reply:
Looks like it will not, as you resize `result` to `resize_`.

Reply:
But, anyway, there is an ambiguity in the shape of the result for `p=2`. This has to be documented.

Reply:
A better idea is to perform all reduction without `keepdim`, then you do not have to do swaps and stuff, and then insert fake dimensions at the very end with `unsqueeze` if `keepdim=True`.

Reply:
The case of `p==2` is peculiar, as it is not obvious which dimension actually contains norms, especially with `keepdim=True`. This requires a proper documentation I presume.

Reply:
You're right that this implementation doesn't give the correct result with keepdim. I added tests for it, and there were several problems with keepdim for nuclear/frobenius/matrix norms. I've fixed all but one of the failing cases now, and I'll push the update once that's done. For matrix norm p=+/-2, my solution was to create a `reverse_permuted_dims` vector so I can run `permute` at the end of the operation to put the dims all back in their original order--after unsqueezing the last dimension, similar to your suggestion.

> then you do not have to do swaps and stuff

Unfortunately, I don't think the permute operation is avoidable. SVD doesn't allow the caller to specify the dims it will operate on, it will always look at the last two dims.